File: app/src/main/java/br/com/jonathanzanella/myexpenses/card/CardPresenter.java
package br.com.jonathanzanella.myexpenses.card;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.annotation.WorkerThread;

import org.joda.time.DateTime;

import java.util.List;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.account.AccountRepository;
import br.com.jonathanzanella.myexpenses.account.ListAccountActivity;
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.ResourcesHelper;
import br.com.jonathanzanella.myexpenses.validations.OperationResult;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;

import static android.app.Activity.RESULT_OK;

class CardPresenter {
	private static final int REQUEST_SELECT_ACCOUNT = 1006;
	private final CardRepository repository;
	private final AccountRepository accountRepository;
	private final ExpenseRepository expenseRepository;
	private final ResourcesHelper resourcesHelper;

	private CardContract.View view;
	@Nullable
	private CardContract.EditView editView;
	private Card card;
	private Account account;

	CardPresenter(CardRepository repository, AccountRepository accountRepository,
	              ExpenseRepository expenseRepository, ResourcesHelper resourcesHelper) {
		this.repository = repository;
		this.accountRepository = accountRepository;
		this.expenseRepository = expenseRepository;
		this.resourcesHelper = resourcesHelper;
	}

	void attachView(CardContract.View view) {
		this.view = view;
	}

	void attachView(CardContract.EditView view) {
		this.view = view;
		this.editView = view;
	}

	void detachView() {
		this.view = null;
		this.editView = null;
	}

	@UiThread
	void viewUpdated(boolean invalidateCache) {
		if (card != null) {
			if(invalidateCache) {
				new AsyncTask<Void, Void, Void>() {

					@Override
					protected Void doInBackground(Void... voids) {
						card = repository.find(card.getUuid());
						return null;
					}

					@Override
					protected void onPostExecute(Void aVoid) {
						super.onPostExecute(aVoid);
						updateView();
					}
				}.execute();
			} else {
				updateView();
			}
		} else {
			updateView();
		}
	}

	@UiThread
	void updateView() {
		if (card != null) {
			if (editView != null) {
				editView.setTitle(R.string.edit_card_title);
			} else {
				String title = view.getContext().getString(R.string.card);
				view.setTitle(title.concat(" ").concat(card.getName()));
			}
			view.showCard(card);
			if(account == null)
				loadAccount(card.getAccountUuid());
			else
				editView.onAccountSelected(account);
		} else {
			if(editView != null)
				editView.setTitle(R.string.new_card_title);
		}
	}

	@WorkerThread
	void loadCard(String uuid) {
		card = repository.find(uuid);
		if(card == null)
			throw new CardNotFoundException(uuid);
	}

	@UiThread
	void save() {
		if(editView == null)
			throw new InvalidMethodCallException("save", getClass().toString(), "View should be a Edit View");
		if(card == null)
			card = new Card(accountRepository);
		card = editView.fillCard(card);
		if(account != null)
			card.setAccount(account);

		new AsyncTask<Void, Void, OperationResult>() {

			@Override
			protected OperationResult doInBackground(Void... voids) {
				return repository.save(card);
			}

			@Override
			protected void onPostExecute(OperationResult result) {
				super.onPostExecute(result);

				if(result.isValid()) {
					editView.finishView();
				} else {
					for (ValidationError validationError : result.getErrors())
						editView.showError(validationError);
				}
			}
		}.execute();
	}

	String getUuid() {
		return card != null ? card.getUuid() : null;
	}

	void showSelectAccountActivity(Activity act) {
		if(card == null)
			act.startActivityForResult(new Intent(act, ListAccountActivity.class), REQUEST_SELECT_ACCOUNT);
	}

	void onActivityResult(int requestCode, int resultCode, Intent data) {
		switch (requestCode) {
			case REQUEST_SELECT_ACCOUNT: {
				if(resultCode == RESULT_OK)
					loadAccount(data.getStringExtra(ListAccountActivity.Companion.getKEY_ACCOUNT_SELECTED_UUID()));
				break;
			}
		}
	}

	@UiThread
	private void loadAccount(final String uuid) {
		new AsyncTask<Void, Void, Account>() {

			@Override
			protected Account doInBackground(Void... voids) {
				return accountRepository.find(uuid);
			}

			@Override
			protected void onPostExecute(Account account) {
				super.onPostExecute(account);
				CardPresenter.this.account = account;
				if(account != null && editView != null) {
					editView.onAccountSelected(account);
				}
			}
		}.execute();
	}

	Expense generateCreditCardBill(DateTime month) {
		List<Expense> expenses = repository.creditCardBills(card, month);
		int totalExpense = 0;
		for (Expense expense : expenses) {
			totalExpense += expense.getValue();
			expense.setCharged(true);
			expenseRepository.save(expense);
		}

		if(totalExpense == 0)
			return null;

		Expense e = new Expense();
		e.setName(resourcesHelper.getString(R.string.invoice) + " " + card.getName());
		e.setValue(totalExpense);
		e.setChargeable(card.getAccount());
		expenseRepository.save(e);

		return e;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/card/CardPresenter.kt
package br.com.jonathanzanella.myexpenses.card

import android.app.Activity
import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.annotation.WorkerThread
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.account.AccountRepository
import br.com.jonathanzanella.myexpenses.account.ListAccountActivity
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException
import br.com.jonathanzanella.myexpenses.exceptions.ValidationException
import br.com.jonathanzanella.myexpenses.expense.Expense
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository
import br.com.jonathanzanella.myexpenses.helpers.ResourcesHelper
import br.com.jonathanzanella.myexpenses.validations.ValidationResult
import org.joda.time.DateTime

class CardPresenter(private val repository: CardRepository, private val accountRepository: AccountRepository,
                     private val expenseRepository: ExpenseRepository, private val resourcesHelper: ResourcesHelper) {

    private var view: CardContract.View? = null
    private var editView: CardContract.EditView? = null
    private var card: Card? = null
    private var account: Account? = null

    fun attachView(view: CardContract.View) {
        this.view = view
    }

    fun attachView(view: CardContract.EditView) {
        this.view = view
        this.editView = view
    }

    fun detachView() {
        this.view = null
        this.editView = null
    }

    @UiThread
    fun viewUpdated(invalidateCache: Boolean) {
        if (card != null) {
            if (invalidateCache) {
                object : AsyncTask<Void, Void, Void>() {

                    override fun doInBackground(vararg voids: Void): Void? {
                        card = repository.find(card!!.uuid!!)
                        return null
                    }

                    override fun onPostExecute(aVoid: Void?) {
                        super.onPostExecute(aVoid)
                        updateView()
                    }
                }.execute()
            } else {
                updateView()
            }
        } else {
            updateView()
        }
    }

    @UiThread
    fun updateView() {
        val c = card
        val v = editView
        if (c != null) {
            if (v != null) {
                v.setTitle(R.string.edit_card_title)
            } else {
                view!!.let {
                    val title = it.context.getString(R.string.card)
                    it.setTitle(title + " " + c.name)
                }
            }
            view!!.showCard(c)
            val a = account
            if (a == null)
                loadAccount(c.accountUuid!!)
            else
                v?.onAccountSelected(a)
        } else {
            v?.setTitle(R.string.new_card_title)
        }
    }

    @UiThread
    fun reloadCard() {
        object : AsyncTask<Void, Void, Void>() {

            override fun doInBackground(vararg voids: Void): Void? {
                loadCard(card!!.uuid!!)
                return null
            }

            override fun onPostExecute(aVoid: Void?) {
                super.onPostExecute(aVoid)
                updateView()
            }
        }.execute()
    }

    @WorkerThread
    fun loadCard(uuid: String) {
        card = repository.find(uuid)
        if (card == null)
            throw CardNotFoundException(uuid)
    }

    @UiThread
    fun save() {
        val v = editView ?: throw InvalidMethodCallException("save", javaClass.toString(), "View should be a Edit View")
        if (card == null)
            card = Card(accountRepository)
        card = v.fillCard(card!!)
        if (account != null)
            card!!.account = account

        object : AsyncTask<Void, Void, ValidationResult>() {

            override fun doInBackground(vararg voids: Void): ValidationResult {
                return repository.save(card!!)
            }

            override fun onPostExecute(result: ValidationResult) {
                super.onPostExecute(result)

                if (result.isValid) {
                    v.finishView()
                } else {
                    for (validationError in result.errors)
                        v.showError(validationError)
                }
            }
        }.execute()
    }

    val uuid: String?
        get() = card?.uuid

    fun showSelectAccountActivity(act: Activity) {
        if (card == null)
            act.startActivityForResult(Intent(act, ListAccountActivity::class.java), REQUEST_SELECT_ACCOUNT)
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        when (requestCode) {
            REQUEST_SELECT_ACCOUNT -> {
                if (resultCode == RESULT_OK)
                    loadAccount(data.getStringExtra(ListAccountActivity.KEY_ACCOUNT_SELECTED_UUID))
            }
        }
    }

    @UiThread
    private fun loadAccount(uuid: String) {
        object : AsyncTask<Void, Void, Account>() {

            override fun doInBackground(vararg voids: Void): Account? {
                return accountRepository.find(uuid)
            }

            override fun onPostExecute(account: Account?) {
                super.onPostExecute(account)
                this@CardPresenter.account = account
                if (account != null) {
                    editView?.onAccountSelected(account)
                }
            }
        }.execute()
    }

    fun generateCreditCardBill(month: DateTime): Expense? {
        val c = card!!
        val expenses = repository.creditCardBills(c, month)
        var totalExpense = 0
        for (expense in expenses) {
            totalExpense += expense.value
            expense.charged = true
            expenseRepository.save(expense)
        }

        if (totalExpense == 0)
            return null

        val e = Expense()
        e.name = resourcesHelper.getString(R.string.invoice) + " " + c.name
        e.setDate(DateTime.now())
        e.value = totalExpense
        e.setChargeable(c.account!!)
        val validationResult = expenseRepository.save(e)
        if (!validationResult.isValid)
            throw ValidationException(validationResult)

        return e
    }

    companion object {
        private val REQUEST_SELECT_ACCOUNT = 1006
    }
}
